pkg/sources/docker/registries.go

```diff
@@ -45,9 +46,72 @@ func MakeRegistryFromNamespace(namespace string) Registry {
 	return registry
 }
 
-// === DockerHub registry ===
+const (
+	maxRetriesOnRateLimit = 5
+	maxSleepTime          = 30 * time.Second
+)
+
+func requestWithRateLimit(ctx context.Context, method, urlStr string, headers http.Header) (*http.Response, error) {
+	for attempts := 0; attempts < maxRetriesOnRateLimit; attempts++ {
+		req, err := http.NewRequestWithContext(ctx, method, urlStr, http.NoBody)
+		if err != nil {
+			return nil, err
+		}
+
+		// Copy headers each time to avoid mutating the caller's map.
+		for k, vs := range headers {
+			for _, v := range vs {
+				req.Header.Add(k, v)
+			}
+		}
+
+		resp, err := defaultHTTPClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode != http.StatusTooManyRequests {
+			return resp, nil
+		}
+
+		// We'll honor Retry-After if it's present, otherwise back off a little.
+		retryAfter := resp.Header.Get("Retry-After")
 
-// DockerHub implements the Registry interface for Docker Hub.
+		_, _ = io.Copy(io.Discard, resp.Body)
+		_ = resp.Body.Close()
+
+		sleepFor := 5 * time.Second // 5 second default backoff
+
+		if retryAfter != "" {
+			// Retry-After can be either seconds or an HTTP date.
+			if secs, err := strconv.Atoi(retryAfter); err == nil && secs >= 0 {
+				sleepFor = time.Duration(secs) * time.Second
+			} else if t, err := http.ParseTime(retryAfter); err == nil {
+				if duration := time.Until(t); duration > 0 {
+					sleepFor = duration
+				}
+			}
+		}
+
+		// Rely on our own max sleep time instead of Retry-After, in case it's too long.
+		if sleepFor > maxSleepTime {
+			sleepFor = maxSleepTime
+		}
+
+		// Respect context cancellation while sleeping.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(sleepFor):
+		}
+	}
+
+	return nil, fmt.Errorf("rate limited (HTTP status 429) after %d attempts", maxRetriesOnRateLimit)
+}
```

Comment:
The logic looks solid here. Just wondering, have you considered exponential backoff?